src/main/java/ch/njol/skript/expressions/ExprElement.java

```diff
@@ -37,79 +38,120 @@
+@Since("2.0, INSERT VERSION (relative to last element, range of elements)")
 public class ExprElement extends SimpleExpression<Object> {
 
+	private static final String objects = " [out] of %objects%";
+	private static final Patterns<ElementType> PATTERNS = new Patterns<>(new Object[][]{
+		{"[the] (first|:last) element" + objects, ElementType.SINGLE},
+		{"[the] (first|:last) %number% elements" + objects, ElementType.MULTIPLE},
+		{"[a] random element" + objects, ElementType.RANDOM},
+		{"[the] %number%(st|nd|rd|th) [last:[to] last] element" + objects, ElementType.ORDINAL},
+		{"[the] elements (from|between) %number% (to|and) %number%" + objects, ElementType.RANGE}
+	});
+
 	static {
-		Skript.registerExpression(ExprElement.class, Object.class, ExpressionType.PROPERTY, "(0:[the] first|1:[the] last|2:[a] random|3:[the] %-number%(st|nd|rd|th)|4:[the] %-number%(st|nd|rd|th) [to] last) element [out] of %objects%");
+		Skript.registerExpression(ExprElement.class, Object.class, ExpressionType.PROPERTY, PATTERNS.getPatterns());
 	}
 
 	private enum ElementType {
-		FIRST, LAST, RANDOM, ORDINAL, TAIL_END_ORDINAL
+		SINGLE, MULTIPLE, RANDOM, ORDINAL, RANGE
 	}
 
-	private ElementType type;
-
 	private Expression<?> expr;
-
-	@Nullable
-	private Expression<Number> number;
+	private	@Nullable Expression<Number> number, to;
+	private ElementType type;
+	private boolean last;
 
 	@Override
 	@SuppressWarnings("unchecked")
 	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
-		expr = LiteralUtils.defendExpression(exprs[2]);
-		type = ElementType.values()[parseResult.mark];
-		number = (Expression<Number>) (type == ElementType.ORDINAL ? exprs[0]: exprs[1]);
+		type = PATTERNS.getInfo(matchedPattern);
+		expr = LiteralUtils.defendExpression(exprs[exprs.length - 1]);
+		switch (type) {
+			case RANGE:
+				to = (Expression<Number>) exprs[1];
+				//$FALL-THROUGH$
+			case MULTIPLE:
+			case ORDINAL:
+				number = (Expression<Number>) exprs[0];
+				break;
+			default:
+				number = null;
+				break;
+		}
+		last = parseResult.hasTag("last");
 		return LiteralUtils.canInitSafely(expr);
 	}
 
 	@Override
 	@Nullable
 	protected Object[] get(Event event) {
-		Iterator<?> iter = expr.iterator(event);
-		if (iter == null || !iter.hasNext())
+		Iterator<?> iterator = expr.iterator(event);
+		if (iterator == null || !iterator.hasNext())
 			return null;
 		Object element = null;
 		switch (type) {
-			case FIRST:
-				element = iter.next();
+			case SINGLE:
+				element = last ? Iterators.getLast(iterator) : iterator.next();
 				break;
-			case LAST:
-				element = Iterators.getLast(iter);
+			case RANDOM:
+				Object[] allIterValues = Iterators.toArray(iterator, Object.class);
+				element = CollectionUtils.getRandom(allIterValues);
 				break;
 			case ORDINAL:
-				assert this.number != null;
+				allIterValues = Iterators.toArray(iterator, Object.class);
```

Comment:
the old implementation had better performance - this one always computes all the values, while the old impl computes as few elements as possible